Tutorial: Get started with k6 (#1192)
grafana/k6-docs: src/data/markdown/translated-guides/en/02 Using k6/00 Tutorial/200 Test for performance.md

---

In the previous section, you made a working script to test the endpoint functionality.
The next step is to test how this system responds under load.
This requires using the powerful [`options`](/using-k6/options) object, which configures the parts of the test that don't deal with user behavior.

In this tutorial, learn how to:
- Use [thresholds](/using-k6/thresholds) to assert for performance criteria
- Configure load increases through [scenarios](/using-k6/scenarios)

These examples build on the script from the previous section.

## Context: meet service-level objectives

To assess the login endpoint's performance, your team may have defined [service level objectives](https://sre.google/sre-book/service-level-objectives/) (SLOs):
- 99.9% of requests are successful
- 99% of requests have a latency of 1000ms or less.

The service must meet these SLOs up to peak traffic of 1 request/second.

After you confirm run the peak traffic test, run another test to determine where the system degrades (this is also often done to discover baselines before the organization has an SLOs).

## Assert for performance with thresholds

> To codify the SLOs, add [_thresholds_](/using-k6/thresholds) to test that your system performs to its goal criteria.
> Thresholds are exported in options.

Reviewer: ```suggestion
To codify the SLOs, add [_thresholds_](/using-k6/thresholds) to test that your system performs to its goal criteria.
Thresholds are set in options.
```

If we expected zero knowledge for js, using `exported` seems like not a great idea given that users will just not know what that means.

Also we request that they copy code so ... hopefully they will do that.

Author: I've tried to make this more explicit and make the example console output cohere in [4375753](https://github.com/grafana/k6-docs/pull/1192/commits/43757539d17e5e3aa74b773122ca692d0effb88a)

```javascript
export const options = {
  //define thresholds
  thresholds: {
    http_req_failed: ['rate<0.01'], // http errors should be less than 1%
    http_req_duration: ['p(99)<200'], // 99% of requests should be below 200ms
  },
};
```

> Add these thresholds object to your script and run it.

Reviewer: ```suggestion
Add this option object with thresholds to your script and run it.
```

I am not certain how explicit this is and if somebody with less experience will not end up copying it inside the `default` function for example :shrug:

Inspect the console output to determine whether performance crossed a threshold.

```
✓ http_req_duration..............: avg=66.14ms    min=0s         med=0s         max=198.42ms   p(90)=158.73ms   p(95)=178.58ms  
       { expected_response:true }...: avg=198.42ms   min=198.42ms   med=198.42ms   max=198.42ms   p(90)=198.42ms   p(95)=198.42ms  

> ✗ http_req_failed................: 66.66% ✓ 2        ✗ 1    

Author: Made it one in [4375753](https://github.com/grafana/k6-docs/pull/1192/commits/43757539d17e5e3aa74b773122ca692d0effb88a)

However, this test server may be under load by many k6 learners, so it's hard to say.

<Blockquote mod="note" title="To visualize results...">

At this point, it'd be nice to have a graphical interface to visualize metrics as they occur.
k6 has many output formats, which can serve as inputs for many visualization tools, both open source and commercial.
For ideas, read [Ways to visualize k6 results](https://k6.io/blog/ways-to-visualize-results).

</Blockquote>

### Ramp up until threshold fails

Finally, run a [stress test](/test-types/stress-testing), where you probe the limits of the system.
In this case, run the test until the availability (error rate) threshold is crossed.

To do this:

1. Configure the threshold so that it aborts when it fails.

  ```json
    http_req_failed: [{ threshold: "rate<0.01", abortOnFail: true }], // availability threshold for error rate
  }
  ```

1. Use the `ramping-arrival-rate` scenario to ramp the test up until it fails.
  This configuration expresses load not in VUs but in iterations per minute.

  ```json
    scenarios: {
      //arbitrary name of scenario:
      stress_test: {
        executor: "ramping-arrival-rate",

>         // Start iterations per `timeUnit`
>         startRate: 20,
>         // Start `startRate` iterations per minute

Reviewer: ```suggestion
        // Start `startRate` iterations per minute
        startRate: 20,
        // Start iterations per `timeUnit`
```

they seem reversed to me ?
